lib/ansible/modules/crypto/luks_device.py

```diff
@@ -0,0 +1,448 @@
+                             % (device, result[STDERR]))
+
+    def run_luks_add_key(self, device, key, new_key):
+        ''' Add 'new_key' to given 'device'; authentization done using 'key'
+            Raises ValueError when command fails
+        '''
+        result = self._run_command(['cryptsetup', 'luksAddKey', device,
+                                    new_key, '--key-file', key])
+        if result[RETURN_CODE] != 0:
+            raise ValueError('Error while adding new LUKS key to %s: %s'
+                             % (device, result[STDERR]))
+
+    def run_luks_remove_key(self, device, key):
+        ''' Remove key from given device
+            Raises ValueError when command fails
+        '''
+        result = self._run_command(['cryptsetup', 'luksRemoveKey', device,
+                                    '-q', '--key-file', key])
+        if result[RETURN_CODE] != 0:
+            raise ValueError('Error while removing LUKS key from %s: %s'
+                             % (device, result[STDERR]))
+
+    def generate_luks_name(self, device):
+        ''' Generate name for luks based on device UUID ('luks-<UUID>').
+            Raises ValueError when obtaining of UUID fails.
+        '''
+        result = self._run_command(['lsblk', '-n', device, '-o', 'UUID'])
+
+        if result[RETURN_CODE] != 0:
+            raise ValueError('Error while generating LUKS name for %s: %s'
+                             % (device, result[STDERR]))
+        dev_uuid = result[STDOUT].strip()
+        return 'luks-%s' % dev_uuid
+
+    def is_luks(self, device):
+        ''' check if the LUKS device does exist
+        '''
+        result = self._run_command(['cryptsetup', 'isLuks', device])
+        return result[RETURN_CODE] == 0
+
+    def get_container_name_by_device(self, device):
+        ''' obtain LUKS container name based on the device where it is located
+            return None if not found
+            raise ValueError if lsblk command fails
+        '''
+        result = self._run_command(['lsblk', device, '-nlo', 'type,name'])
+        if result[RETURN_CODE] != 0:
+            raise ValueError('Error while obtaining LUKS name for %s: %s'
+                             % (device, result[STDERR]))
+
+        m = LUKS_NAME_REGEX.search(result[STDOUT])
+
+        try:
+            name = m.group(1)
+        except AttributeError as e:
+            name = None
+        return name
+
+    def get_container_device_by_name(self, name):
+        ''' obtain device name based on the LUKS container name
+            return None if not found
+            raise ValueError if lsblk command fails
+        '''
+        # apparently each device can have only one LUKS container on it
+        result = self._run_command(['cryptsetup', 'status', name])
+        if result[RETURN_CODE] != 0:
+            return None
+
+        m = LUKS_DEVICE_REGEX.search(result[STDOUT])
+        device = m.group(1)
+        return device
+
+
+class Conditions(CryptHandler):
```

Comment:
Instead of deriving from `CryptHandler`, why not pass `CryptHandler` as a constructor argument?

Otherwise, I don't see why you need both a `CryptHandler` and a `Conditions` instance in the main module code (since you can use the `Conditions` instance for both).

Reply:
Those two classes do different things and need to be distinguished. However, I agree that this solution is not optimal. To address this I did a bit of rework: Now there are three classes:
`Handler` - basic class that takes care of generic stuff both of the former classes needed.
`CryptHandler` - inherits from `Handler`, generic stuff was moved to it from here.
`ConditionsHandler` - inherits from `Handler`, what was formerly needed from old `CryptHandler` is now moved to `Handler`.

I think that this approach is much more elegant and simplifies understanding of the workings of the module.

Reply:
Yes, that's a better solution.

BTW, why not put all functionality calling `cryptsetup` into `CryptHandler`, and giving `ConditionsHandler` a reference to `CryptHandler` for the calls it needs to do? Then there's a more clear separation into `CryptHandler`, which does all the `cryptsetup` calls, and `ConditionsHandler`, which contains most of the decision logic.

Reply:
Fixed.